spirv_msl.cpp

```diff
@@ -4422,6 +4645,59 @@ bool CompilerMSL::skip_argument(uint32_t) const
 	return false;
 }
 
+void CompilerMSL::analyze_image_and_sampler_usage()
+{
+	Compiler::analyze_image_and_sampler_usage();
+	if (msl_options.swizzle_texture_samples)
+	{
+		SampledImageScanner scanner(*this);
+		traverse_all_reachable_opcodes(get<SPIRFunction>(entry_point), scanner);
+	}
+}
+
+bool CompilerMSL::SampledImageScanner::handle(spv::Op opcode, const uint32_t *args, uint32_t length)
+{
+	switch (opcode)
+	{
+	case OpLoad:
+	case OpImage:
+	case OpSampledImage:
+	{
+		if (length < 3)
+			return false;
+
+		uint32_t result_type = args[0];
+		auto &type = compiler.get<SPIRType>(result_type);
+		if ((type.basetype != SPIRType::Image && type.basetype != SPIRType::SampledImage) || type.image.sampled != 1)
+			return true;
+
+		uint32_t id = args[1];
+		compiler.set<SPIRExpression>(id, "", result_type, true);
+		break;
+	}
+	case OpImageSampleExplicitLod:
+	case OpImageSampleProjExplicitLod:
+	case OpImageSampleDrefExplicitLod:
+	case OpImageSampleProjDrefExplicitLod:
+	case OpImageSampleImplicitLod:
+	case OpImageSampleProjImplicitLod:
+	case OpImageSampleDrefImplicitLod:
+	case OpImageSampleProjDrefImplicitLod:
+	case OpImageFetch:
+	case OpImageGather:
+		compiler.has_sampled_images |= is_sampled_image_type(compiler.expression_type(args[2]));
+		break;
+	default:
+		break;
+	}
+	return true;
+}
+
+bool CompilerMSL::SampledImageScanner::is_sampled_image_type(const SPIRType &type)
```

Comment:
This seems generally useful for other paths in this patch, so might make sense to hoist it out to Compiler.

Reply:
Should probably add a check for type.basetype == SPIRType::Image || type.basetype == SPIRType::SampledImage in that case to make it general.